packages/dialect-react-ui/components/Notifications/EmailForm.tsx

```diff
@@ -0,0 +1,222 @@
+
+  const isEmailSaved = Boolean(emailObj);
+  const isChanging = emailObj && isEmailEditing;
+  const isVerified = emailObj?.verified;
+
+  const currentError =
+    emailError ||
+    fetchingAddressesError ||
+    savingAddressError ||
+    deletingAddressError;
+
+  useEffect(() => {
+    // Update state if addresses updated
+    setEnabled(Boolean(emailObj?.enabled));
+    setEmail(emailObj?.value || '');
+    setEmailEditing(!emailObj?.enabled);
+  }, [emailObj]);
+
+  return (
+    <div>
+      <ValueRow className="mb-2" label="Enable email notifications">
+        <Toggle
+          type="checkbox"
+          checked={isEnabled}
+          onClick={async () => {
+            const nextValue = !isEnabled;
+            if (emailObj && emailObj.enabled !== nextValue) {
+              // TODO: handle error
+              await updateAddress(wallet, {
+                id: emailObj.id,
+                enabled: nextValue,
+              });
+            }
+            setEnabled(!isEnabled);
+          }}
+        />
+      </ValueRow>
+      {isEnabled && (
+        <form onSubmit={(e) => e.preventDefault()}>
+          <div className="flex flex-col space-y-2 mb-2">
+            <div className="">
+              {isEmailSaved && !isEmailEditing ? (
+                <div
+                  className={cs(highlighted, textStyles.body, colors.highlight)}
+                >
+                  <span className="opacity-40">🔗 Email submitted</span>
+                </div>
+              ) : (
+                <input
+                  className={cs(
+                    outlinedInput,
+                    emailError && '!border-red-500 !text-red-500',
+                    'w-full basis-full'
+                  )}
+                  placeholder="Enter email"
+                  type="email"
+                  value={email}
+                  onChange={(e) => setEmail(e.target.value)}
+                  onBlur={(e) =>
+                    e.target.checkValidity()
+                      ? setEmailError(null)
+                      : setEmailError(incorrectEmail)
+                  }
+                  onInvalid={(e) => {
+                    e.preventDefault();
+                    setEmailError(incorrectEmail);
+                  }}
+                  pattern="^\S+@\S+\.\S+$"
+                  disabled={isEmailSaved && !isEmailEditing}
+                />
+              )}
+            </div>
+
+            {isChanging && (
+              <div className="flex flex-row space-x-2">
+                <Button
+                  defaultStyle={secondaryButton}
+                  className="basis-1/2"
+                  onClick={() => {
+                    setEmailEditing(false);
+                  }}
+                >
+                  Cancel
+                </Button>
+                <Button
+                  className={'basis-1/2'}
+                  disabled={email === ''}
+                  onClick={async () => {
```

Comment:
I'd recommend moving this function to a separate variable above, no need to wrap it with useCallback, since this component rerenders every time we type something inside our input.

Reply:
done